relayer/crates/starknet-chain-components/src/types/cosmos/update.rs

```diff
@@ -1,47 +1,881 @@
+        let product![seconds, nanos] = encoding.decode_mut(buffer)?;
+        Ok(ProtoTimestamp { seconds, nanos })
+    }
+}
+
+pub struct EncodeCommitBlockIdFlag;
+
+delegate_components! {
+    EncodeCommitBlockIdFlag {
+        [
+            MutEncoderComponent,
+            MutDecoderComponent,
+        ]: EncodeVariantFrom<Self>,
+    }
+}
+
+impl TransformerRef for EncodeCommitBlockIdFlag {
+    type From = BlockIdFlag;
+    type To<'a> = Sum![(), (), (), ()];
+
+    fn transform<'a>(from: &'a BlockIdFlag) -> Self::To<'a> {
+        match from {
+            BlockIdFlag::Absent => Either::Right(Either::Left(())),
+            BlockIdFlag::Commit => Either::Right(Either::Right(Either::Left(()))),
+            BlockIdFlag::Nil => Either::Right(Either::Right(Either::Right(Either::Left(())))),
+        }
+    }
+}
+
+impl Transformer for EncodeCommitBlockIdFlag {
+    type From = Sum![(), (), (), ()];
+    type To = BlockIdFlag;
+
+    fn transform(from: Self::From) -> Self::To {
+        match from {
+            Either::Left(()) => unreachable!(),
+            Either::Right(Either::Left(())) => BlockIdFlag::Absent,
+            Either::Right(Either::Right(Either::Left(()))) => BlockIdFlag::Commit,
+            Either::Right(Either::Right(Either::Right(Either::Left(())))) => BlockIdFlag::Nil,
+            Either::Right(Either::Right(Either::Right(Either::Right(v)))) => match v {},
+        }
+    }
+}
+
+pub struct EncodeCommitSig;
+
+#[cgp_provider(MutEncoderComponent)]
+impl<Encoding, Strategy> MutEncoder<Encoding, Strategy, CommitSig> for EncodeCommitSig
+where
+    Encoding: CanEncodeMut<Strategy, Product![BlockIdFlag, account::Id, ProtoTimestamp, Signature]>
+        + CanEncodeMut<Strategy, Product![BlockIdFlag, Vec<u8>, ProtoTimestamp, Vec<u8>]>,
+{
+    fn encode_mut(
+        encoding: &Encoding,
+        value: &CommitSig,
+        buffer: &mut Encoding::EncodeBuffer,
+    ) -> Result<(), Encoding::Error> {
+        match value {
+            CommitSig::BlockIdFlagAbsent => encoding.encode_mut(
+                &product![BlockIdFlag::Absent, Vec::new(), ZERO_TIMESTAMP, Vec::new()],
+                buffer,
+            )?,
+            CommitSig::BlockIdFlagCommit {
+                validator_address,
+                timestamp,
+                signature,
+            } => encoding.encode_mut(
+                &product![
+                    BlockIdFlag::Commit,
+                    *validator_address,
+                    (*timestamp).into(),
+                    signature.clone().unwrap()
+                ],
+                buffer,
+            )?,
+            CommitSig::BlockIdFlagNil {
+                validator_address,
+                timestamp,
+                signature,
+            } => encoding.encode_mut(
+                &product![
+                    BlockIdFlag::Nil,
+                    *validator_address,
+                    (*timestamp).into(),
+                    signature.clone().unwrap()
```

Comment:
Same here, can we handle `Option<T>` with proper error?

Reply:
540312f